Userland/Libraries/LibJS/Runtime/FinalizationRegistry.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2021, Idan Horowitz <idan.horowitz@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibJS/Runtime/FinalizationRegistry.h>
+
+namespace JS {
+
+FinalizationRegistry* FinalizationRegistry::create(GlobalObject& global_object, Function& cleanup_callback)
+{
+    return global_object.heap().allocate<FinalizationRegistry>(global_object, *global_object.finalization_registry_prototype(), cleanup_callback);
+}
+
+FinalizationRegistry::FinalizationRegistry(Object& prototype, Function& cleanup_callback)
+    : Object(prototype)
+    , WeakContainer(heap())
+    , m_cleanup_callback(&cleanup_callback)
+{
+}
+
+FinalizationRegistry::~FinalizationRegistry()
+{
+}
+
+void FinalizationRegistry::add_finalization_record(Cell& target, Value held_value, Object* unregister_token)
+{
+    VERIFY(!held_value.is_empty());
+    m_records.append({ &target, held_value, unregister_token });
+}
+
+bool FinalizationRegistry::remove_by_token(Object& unregister_token)
+{
+    auto removed = false;
+    for (auto it = m_records.begin(); it != m_records.end(); ++it) {
+        if (it->unregister_token == &unregister_token) {
+            m_records.remove(it);
+            removed = true;
+        }
+    }
+    return removed;
+}
+
+void FinalizationRegistry::remove_sweeped_cells(Badge<Heap>, Vector<Cell*>& cells)
+{
+    auto any_cells_were_sweeped = false;
+    for (auto cell : cells) {
+        for (auto& record : m_records) {
+            if (record.target != cell)
+                continue;
+            record.target = nullptr;
+            any_cells_were_sweeped = true;
+            break;
+        }
+    }
+    if (any_cells_were_sweeped)
+        vm().enqueue_finalization_registry_cleanup_job(*this);
+}
+
+// 9.13 CleanupFinalizationRegistry ( finalizationRegistry ), https://tc39.es/ecma262/#sec-cleanup-finalization-registry
+void FinalizationRegistry::cleanup(Function* callback)
+{
+    auto cleanup_callback = callback ?: m_cleanup_callback;
+    for (auto it = m_records.begin(); it != m_records.end(); ++it) {
+        if (it->target != nullptr)
+            continue;
+        (void)vm().call(*cleanup_callback, js_undefined(), it->held_value);
+        m_records.remove(it);
+        if (vm().exception())
+            return;
```

Comment:
Let's keep a reference to the VM at the beginning of the function as usual (`auto& vm = this->vm()`) - `vm.foo()` reads a bit nicer later on :)
(and technically slightly more performant, although I trust the compiler to do the right thing)

Reply:
done :)